---
output: github_document
---
<!--
The README should be used to describe the program. It acts like the homepage of
your module.

Edit README.Rmd not README.md. The .Rmd file can be knitted to parse real-code
examples and show their output in the .md file.

To knit, use devtools::build_readme() or outsider.devtools::build()

Edit the template to describe your program: how to install, import and run;
run exemplary, small demonstrations; present key arguments; provide links and
references to the program that the module wraps.

Learn more about markdown and Rmarkdown:
https://daringfireball.net/projects/markdown/syntax
https://rmarkdown.rstudio.com/
-->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# Run [`pasta`](https://github.com/smirarab/pasta) with `outsider` in R

[![Build Status](https://travis-ci.org/dombennett/om..pasta.svg?branch=master)](https://travis-ci.org/dombennett/om..pasta)

> pasta (Practical Alignment using SATe and Transitivity): Multiple sequence
alignment of biological sequences.


<!-- Install information -->
## Install and look up help

```{r install-snippet, eval=TRUE, include=TRUE}
library(outsider)
module_install(repo = "dombennett/om..pasta")
#module_help(repo = "dombennett/om..pasta")
```

<!-- Detailed examples -->
## Aligning a small set of DNA sequences

> The below example is adapted from the
[PASTA tutorial](https://github.com/smirarab/pasta/blob/master/pasta-doc/pasta-tutorial.md#using-pasta).

```{r detailed-example, eval=TRUE, include=TRUE}
library(outsider)
# import
pasta <- module_import('pasta', repo = 'dombennett/om..pasta')
# get help
pasta('-h')

# download
wd <- file.path(tempdir(), 'example_job')
if (!dir.exists(wd)) {
  dir.create(wd)
}
# example DNA, based on MAFFT example
seq_file <- file.path(wd, 'example_seq.fasta')
url <- 'https://raw.githubusercontent.com/DomBennett/om..pasta/master/example_seq.fasta'
download.file(url = url, destfile = seq_file)
# run DNA alignment
pasta(arglist = c('-i', seq_file, '-d', 'dna',
                  '--alignment-suffix=alignment.fasta',
                  '--job=example'), outdir = wd)

# Verify results
list.files(wd)
```

The final alignment is contained within the "example.alignment.fasta" file.

<!-- Remove module after running above example -->
```{r uninstall-snippet, eval=TRUE, include=FALSE}
if ('wd' %in% ls() && dir.exists(wd)) {
  unlink(x = wd, recursive = TRUE, force = TRUE)
}
module_uninstall(repo = 'dombennett/om..pasta')
```

### Key arguments

All PASTA arguments can be provided to the `arglist` as a character
vector. Below is a table describing some of the most common. All arguments can
be displayed using the `-h` flag.

|Argument|Usage|Description|
|--------|-----|-----------|
|i       |-i `file`|Specify input sequence file|
|t       |-t `file`|Specify starting tree file|
|d    |-d `type`|Specify sequence type (protein, RNA, DNA)|
|alignment suffix|--alignment-suffix=`name`|Specify the word ending for the resulting alignment|
|job|--job=`name`|Specify a name for the job|

In addition to the arguments provided in the `arglist` a user can provide an
`outdir` specifying the file location of where all resulting PASTA analysis
files should be returned.

## Links

Find out more by visiting
[PASTA's homepage](https://github.com/smirarab/pasta#options).

## Please cite

* Mirarab S, Nguyen N, Warnow T. PASTA: ultra-large multiple sequence
alignment. Sharan R, ed. Res Comput Mol Biol. 2014:177-191.
* Mirarab S, Nguyen N, Guo S, Wang L-S, Kim J, Warnow T. PASTA: Ultra-Large
Multiple Sequence Alignment for Nucleotide and Amino-Acid Sequences. J Comput
Biol. 2015;22(5):377-386. doi:10.1089/cmb.2014.0156.
* Bennett et al. (2020). outsider: Install and run programs, outside of R,
inside of R. *Journal of Open Source Software*, In review


<!-- Footer -->
---

<img align="left" width="120" height="125" src="https://raw.githubusercontent.com/AntonelliLab/outsider/master/logo.png">

**An `outsider` module**

Learn more at [outsider website](https://antonellilab.github.io/outsider/).
Want to build your own module? Check out [`outsider.devtools` website](https://antonellilab.github.io/outsider.devtools/).
